### ISSUE TITLE
Push rejected, failed to compile Node.js app. Meteor 1.2

### ISSUE THREAD

**Issue by jsantana90** [AdmitHub/meteor-buildpack-horse#132]
I'm having an issue since today when I try to `git push heroku master` a Meteor 1.2 version app

These are the logs from heroku:

```
-----> Node.js app detected
-----> Installing meteor
Downloading Meteor distribution
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
Retrying download in 5 seconds...
 !     Push rejected, failed to compile Node.js app.
 !     Push failed
```

I can't update my app anymore :(

**Comment by jsantana90** [comment]
Here's my `.meteor/release` `METEOR@1.2.2-faster-rebuilds.0`

**Comment by yourcelf** [comment]
The issue here is that Meteor is no longer hosting some file required for the installation.  I saw this behavior with 1.4-beta.7 after 1.4 was released, but didn't expect it to crop up for older non-beta releases.  I had tested with 1.1.0.3 and 1.2.1 and had no problems...

I've opened meteor/meteor#7528 for this -- it's an upstream CDN problem with meteor.  We're encountering this now because we just switched to using the `?release=<VERSION>` strategy to speed up deploys.

In the mean time, you can pin the buildpack to a commit prior to this change to restore the old behavior:

    heroku buildpacks:set https://github.com/AdmitHub/meteor-buildpack-horse.git#5b8fc9280e35ac69f1a14207fd2e97b5590dd47d

**Comment by jsantana90** [comment]
Hi @yourcelf thanks for the quick response!

Unfortunately I wasn't able to deploy successfully using that buildpack.

First case I just unset the normal buildpack and added the one you provided. After that I pushed to heroku but 35 minutes it was still 'building':

```
"/app/tmp/buildpacks/94f277e780091463220fab05e119dc2b407d7d5fc14f4e99949cd0f6971a27a8cac966e395cc4b997229f253d90a749b1a722fb50e37c6b0daf83b9aaf11490a/meteor-WieU/.meteor/packages/meteor-tool/1.4.0-1/mt-os.linux.x86_64/scripts/admin/launch-meteor" /usr/bin/meteor
Then to get started, take a look at 'meteor --help' or see the docs at
docs.meteor.com.
-----> App targets METEOR@1.2.2-faster-rebuilds.0
-----> Bundling bundle
-----> Checking if this meteor version supports --server-only
-----> Attempting to remove android platform.
While removing platforms:
error: android: platform is not in this project
-----> Moving on.
3.10.5
-----> Using 'meteor node', with version: v4.4.7
-----> Building Meteor with ROOT_URL: https://conwastetogo.herokuapp.com
```

Second case, I downgraded my app to meteor 1.2.1 and pushed to heroku, it builded but returned over 300M so I couldn't deploy to heroku.

Third case, I removed the build pack you provided and left the one in master's branch and it deployed successfully.

Using:
Meteor 1.2.1
https://github.com/AdmitHub/meteor-buildpack-horse.git

Works for now.

**Comment by yourcelf** [comment]
Great, I'm happy you got it to work with 1.2.1.  If you needed to stick with 1.2.2-faster-rebuilds.0, further downgrading of buildpack might help with the slug size issue, by going back to a version that didn't try to include the meteor distribution in the built slug:

     heroku buildpacks:set https://github.com/AdmitHub/meteor-buildpack-horse#9249b73808c99197f2c4176fecf29fad92724f14

But not sure if that would help with the build timeout issue.  I'm hoping we can get some response from upstream Meteor on the CDN issues -- if not, we can find and tag an older version of this buildpack that works with older Meteor's that have busted CDN's.

**Comment by jsantana90** [comment]
Bad news, I don't know what happen but after deploying for a second time I got these:

```
-----> Node.js app detected
-----> Meteor version: Meteor 1.4.0.1
-----> Checking if this meteor version supports --server-only
-----> Attempting to remove android platform.
While removing platforms:
error: android: platform is not in this project
-----> Moving on.
cp: cannot stat ‘/app/tmp/cache/1.2.1/.meteor/packages/meteor-tool/.1.4.0-1.lxxk89++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node\n/app/tmp/cache/1.2.1/.meteor/packages/meteor-tool/.1.1.10.ki0ccv++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node’: No such file or directory
 !     Push rejected, failed to compile Node.js app.
 !     Push failed
```

**Comment by yourcelf** [comment]
This is very strange:

`-----> Meteor version: Meteor 1.4.0.1` and the `.1.4.0-1` in the cp path imply that your `.meteor/release` is 1.4.0.1.  However, the directory `/app/tmp/cache/1.2.1/` implies that your meteor release is 1.2.1.

It looks as though somehow Meteor v1.4.0.1 got installed even though you requested v1.2.1.  Could you check the value of `.meteor/release`?

If it is `METEOR@1.2.1`, can you please set an environment variable to clear the cache so that meteor gets reinstalled?

    heroku config:set BUILDPACK_CLEAR_CACHE=1

I wonder if `install.meteor.com/?release=<VERSION>` is unreliable, and occasionally delivering a different version than requested.  Trying to think if there's any other way this could have happened.

**Comment by yourcelf** [comment]
Just to note: I haven't been able to reproduce that outcome yet on a test app targeting 1.2.1.  Meteor v1.2.1 was successfully installed, and subsequent builds work correctly with the cached version.

**Comment by jsantana90** [comment]
Adding `BUILDPACK_CLEAR_CACHE` worked and now I can deploy successfully. Should I leave this environment variable set? or am I safe to take it out?

**Comment by yourcelf** [comment]
You *should* be able to unset it now. Though I still don't know how the cache got into the state it was in -- it looks like somehow you had both 1.4.0.1 and 1.1.10 installed where there should have only been v1.2.1.

`heroku config:unset BUILDPACK_CLEAR_CACHE` will make subsequent builds that still target 1.2.1 faster.

**Comment by gregory** [comment]
`heroku config:set BUILDPACK_CLEAR_CACHE=1` did the trick for me too.

**Comment by zorikcherfas** [comment]
heroku config:set BUILDPACK_CLEAR_CACHE=1 did the trick for me too.

**Comment by yourcelf** [comment]
@gregory and @zorikcherfas: could you please share what version of Meteor you're using, and the output of the build that failed before clearing the cache?  I'm still trying to track down the cause of the errors.  Setting `BUILDPACK_CLEAR_CACHE` is a workaround, but we want a fix.

**Comment by TSDrrodriguez** [comment]
Having the same issue with 1.3.5 as target, the clear cache command also did it for me.

**Comment by jsantana90** [comment]
Can't seem to deploy without the `BUILDPACK_CLEAR_CACHE` trick

```
remote: Building source:
remote: 
remote: -----> Node.js app detected
remote: -----> Meteor version: Meteor 1.4.0.1
remote: -----> Checking if this meteor version supports --server-only
remote: cp: cannot stat ‘/app/tmp/cache/1.3.3.1/.meteor/packages/meteor-tool/.1.4.0-1.1y1hz1n++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node\n/app/tmp/cache/1.3.3.1/.meteor/packages/meteor-tool/.1.3.3_1.9tltar++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node’: No such file or directory
remote:  !     Push rejected, failed to compile Node.js app.
remote: 
remote:  !     Push failed
remote: Verifying deploy....
remote: 
remote: !	Push rejected to myapp.
```

`METEOR@1.3.3.1` with master branch buildpack

**Comment by Philip-Nunoo** [comment]
@jsantana90 heroku config:set BUILDPACK_CLEAR_CACHE=1 should fix it

**Comment by jsantana90** [comment]
@Philip-Nunoo I know, but as @yourcelf stated, this is just a workaround, it should, by default, work without it.

**Comment by rfines** [comment]
I had the same problem 
`heroku config:set BUILDPACK_CLEAR_CACHE=1` 
worked for me. I am using version 1.1.0.3.

```
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Node.js app detected
remote: -----> Meteor version: Meteor 1.4.0.1
remote: -----> Checking if this meteor version supports --server-only
remote: -----> Attempting to remove android platform.
remote: android: platform is not in this project
remote: -----> Moving on.
remote: cp: cannot stat ‘/app/tmp/cache/1.1.0.3/.meteor/packages/meteor-tool/.1.4.0-1.1m5fbp5++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node\n/app/tmp/cache/1.1.0.3/.meteor/packages/meteor-tool/.1.1.4.xvre8c++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node’: No such file or directory
remote:  !     Push rejected, failed to compile Node.js app.
remote:
remote:  !     Push failed
remote: Verifying deploy...
remote:
remote: !	Push rejected
```

**Comment by RobertLowe** [comment]
I'm not sure but: @yourcelf I think the current compile script won't? support 1.2 as `meteor node` or `meteor npm` weren't added until `1.3`/`1.4`, previously node had been installed directly and not used the binaries from the meteor bundle (which is fine now), just two cents on this issue 👍

**Comment by yourcelf** [comment]
@RobertLowe, we are not using `meteor node` or `meteor npm`.  Instead, we're searching within the meteor bundle for node and npm.  This is necessary not only to support older meteor's, but also so that we can copy the node binary into the built slug for space-efficient execution of node.

This buildpack *does* have full support for meteor from 1.0 to 1.4.  This issue concerns a problem with meteor deciding to upgrade itself without us intentionally asking it to; and I still don't know why.   Currently, the reliable workaround is to `heroku config:set BUILDPACK_CLEAR_CACHE=1`, which clears existing meteor installations at the start of every build.  This increases build time a bit, but works around any unwanted upgrades meteor throws at us.

**Comment by RobertLowe** [comment]
Ah, thanks for clarifying, I totally scanned the `compile` source too quickly. Copying is a good idea 👍

**Comment by aexonicarindam** [comment]
Can not upload code in heroku.

Counting objects: 1331, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (1132/1132), done.
Writing objects: 100% (1331/1331), 13.77 MiB | 54.00 KiB/s, done.
Total 1331 (delta 814), reused 357 (delta 118)
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Node.js app detected
remote: -----> Listing HEROKU_ environment variables
remote: HEROKU_APP_ID:           
remote: HEROKU_APP_NAME:         
remote: HEROKU_DYNO_ID:          
remote: HEROKU_RELEASE_VERSION:  
remote: HEROKU_SLUG_COMMIT:      
remote: HEROKU_SLUG_DESCRIPTION: 
remote: ----> Clearing cache dir.
remote: -----> Installing meteor
remote: curl: (3) Illegal characters found in URL
remote: /app/tmp/buildpacks/ec95e5d1b974435c5c7157ba6fcb0d6323f4e70beb6f90c5dc8e2248948ff950dfcd73ff00896ded7938ceb333a80d48f2db8558e376c5c760eb8d9eb9b681d3/bin/compile: 1: /app/tmp/buildpacks/ec95e5d1b974435c5c7157ba6fcb0d6323f4e70beb6f90c5dc8e2248948ff950dfcd73ff00896ded7938ceb333a80d48f2db8558e376c5c760eb8d9eb9b681dremote: /.meteor/meteor: not found
remote: -----> Meteor version: 
remote: -----> Checking if this meteor version supports --server-only
remote: /app/tmp/buildpacks/ec95e5d1b974435c5c7157ba6fcb0d6323f4e70beb6f90c5dc8e2248948ff950dfcd73ff00896ded7938ceb333a80d48f2db8558e376c5c760eb8d9eb9b681d3/bin/compile: 1: /app/tmp/buildpacks/ec95e5d1b974435c5c7157ba6fcb0d6323f4e70beb6f90c5dc8e2248948ff950dfcd73ff00896ded7938ceb333a80d48f2db8558e376c5c760eb8d9eb9b681dremote: /.meteor/meteor: not found
remote: -----> Attempting to remove android platform.
remote: /app/tmp/buildpacks/ec95e5d1b974435c5c7157ba6fcb0d6323f4e70beb6f90c5dc8e2248948ff950dfcd73ff00896ded7938ceb333a80d48f2db8558e376c5c760eb8d9eb9b681d3/bin/compile: 135: /app/tmp/buildpacks/ec95e5d1b974435c5c7157ba6fcb0d6323f4e70beb6f90c5dc8e2248948ff950dfcd73ff00896ded7938ceb333a80d48f2db8558e376c5c760eb8d9eb9b68remote: /.meteor/meteor: not found
remote: -----> Moving on.
remote: FATAL: Can't find npm/node within meteor bundle. This is a bug. Please open an issue at https://github.com/AdmitHub/meteor-buildpack-horse.
remote:  !     Push rejected, failed to compile Node.js app.
remote: 
remote:  !     Push failed
remote: Verifying deploy...
remote: 
remote: !	Push rejected to qup-release6.
remote: 
To https://git.heroku.com/qup-release6.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'https://git.heroku.com/qup-release6.git'

**Comment by yourcelf** [comment]
@aexonicarindam, your issue looks like a different one.  Please open a new issue.

**Comment by bmustata** [comment]
Temporary fix: heroku config:set BUILDPACK_CLEAR_CACHE=1

**Comment by yourcelf** [comment]
Please try the beta release of meteor-buildpack-horse, which should fix problems with locating node/npm within the build, no longer relies on `install.meteor.com/?release=<version>` (which is unreliable).  It also fixes the cache usage so that we don't rely on the version of meteor in the cache being identical to the target release in `.meteor/release`.

```
heroku config:unset BUILDPACK_CLEAR_CACHE
heroku buildpacks:set https://github.com/AdmitHub/meteor-buildpack-horse.git#beta
```

**Comment by fernandobalm** [comment]
My release file has METEOR@1.3.2.4 and https://github.com/AdmitHub/meteor-buildpack-horse.git#beta worked for me on heroku build and deploy.
Thanks!

**Comment by iamfeek** [comment]
Hello!

TLDR; `heroku config:set BUILDPACK_CLEAR_CACHE=1` worked for me.

Meteor version: 1.4.1.1.
Error Message BEFORE Fix: Push rejected, failed to compile Node.js app.
Brief explanation: First deployment worked as expected. After modifying the project, REdeployment produced that error. Running the command once and pushed again, worked.

**Comment by jsantana90** [comment]
Seems like the errors are gone after using the #beta branch, is it safe to leave it with this buildpack?

**Comment by danielflippance** [comment]
The #beta buildpack worked for me. Will I need to replace this with a new non-beta buildpack?

**Comment by yourcelf** [comment]
@danielflippance I'll be merging #beta into master real soon now, as soon as I pass a looming deadline that would prevent me from responding to late breaking issues that people discover as quickly as I'd like.

Once that's done, we'll use the #beta branch again in the future to do bleeding-edge changes that have a higher probability of breaking things.  We try to keep #master as stable as possible.  (We also use #devel for internal testing -- we'll never put something in #beta unless it passes for all of our test repos).

If you'd like to use #beta now and make sure it doesn't change from under you, you're welcome to fork it to your own repo.  Then if changes in meteor or heroku prevent it from working in the future, you can always merge in our changes here.

**Comment by rurri** [comment]
`heroku config:set BUILDPACK_CLEAR_CACHE=1` fixed the same issue. Thanks!

**Comment by greenermoose** [comment]
Same thing here. My app works with meteor 1.4.2.1. heroku build started failing with message:

Meteor version: Meteor 1.4.2.3
Checking if this meteor version supports --server-only
cp: cannot stat '/app/tmp/cache/1.4.2.1/.meteor [...] /bin/node': No such file or directory
Push rejected, failed to compile Node.js app.
Push failed

Fixed with:

heroku config:set BUILDPACK_CLEAR_CACHE=1

**Comment by yourcelf** [comment]
The beta branch has now been merged into master.  It should fix all of these existing problems with cache clearing and node/npm versions.  If you needed `BUILDPACK_CLEAR_CACHE=1` before, you could try unsetting that for slightly faster builds now.
